### ISSUE TITLE
Problems with display of rotational parameters in parameter dialog

### ISSUE THREAD

**Issue by abuntrock** [OpenModelica/OpenModelica#12003]
### Description

Thanks to #7491 the rounding problems, when the displayUnit does not equal the SI unit, are fixed in the current OMEdit nightly. However, there are still problems with the display of rotational parameters, e.g. with the type Modelica.Units.SI.AngularVelocity.

### Steps to Reproduce

1. Create the following modelica class:
```modelica
model DisplayUnitRot

parameter Modelica.Units.SI.AngularVelocity rotVel1(displayUnit="rpm") = 0.10471975511966;
    
parameter Modelica.Units.SI.AngularVelocity rotVel2(displayUnit="rpm") = 104.7197551196598;

end DisplayUnitRot;
```
2. Create an instance of this class and open its parameter dialog in the current nightly and OMEdit 1.22.1.
3. Change the displayUnit from `rpm` to `rad/s` for both parameters in both OMEdit versions. Screenshots for the four different parameters dialogs (two verions times two display units) are attached.
4. Looking at the parameter dialogs with displayUnit changed to `rad/s` reveals an error in the current nightly from our understanding. OMEdit shows the displays the value of `rotVel1` as `0.10471975511966` which is exactly the value listed in the text view. But the current nightly displays `rotVel1` as `0.10471975511966002`. From my naive understanding OMEdit should just display the value that is listed in the text view (which it does in version 1.22.1) but in the current nightly it seems that there is some calculation behind the scenes. It also seems peculiar to me why this does not happen with parameter `rotVel2`, which is displayed the same (and equal to the text view) in both OMEdit versions. Maybe this is due to the fact that the parameter representation of `rotVel2` includes one more digit in the text view. 
5. Looking at the parameter dialogs with the displayUnit equal to `rpm` reveals a question for us. In the old OMEdit 1.22.1, the value of  `rotVel2` is displayed as `1000 rpm`, which is our desired behavior (we want "integer" values as default). But in the new nightly the displayed value of `rotVel2` is `1000.00....1 rpm`. Did the change to the Ryu library lead to a change in significant digits, which can explain this behavior? Then we could add more significant digits to the default value in the text view.

### Expected Behavior

When the parameter's displayUnit equals the SI unit no new digits are added to the end by OMEdit.

### Question
Did the number of significant digits changed as a result of the Ryu library integrating for rounding.

### Screenshots
My collegue @niklwors will add the screenshots as as comment.

### Version and OS

 - OpenModelica Version: 1.23.0-dev-316-g3a47d5ed12 and 1.22.1
 - OS: Windows 10, 64 bit
 - MSL 4.0.0

Thanks for looking into this.
Aaron Buntrock - Bosch Rexroth

**Comment by niklwors** [comment]
picture:
![image001](https://github.com/OpenModelica/OpenModelica/assets/12269776/10783249-0f54-4a44-9d31-a345d6b62e06)

**Comment by abuntrock** [comment]
The problem with `rotVel2` could be solved by increasing the digits written to the text file. Then, its shown as a whole value in the parameter dialog again. It would still be interesting to know if this is a known change due to the usage of the Ryu library.

**Comment by casella** [comment]
The question is, how were those two values obtained in the first place. The first one has 14 significant decimal digits (which is not enough to completely specify a double-precision number, which needs 16), while the second one indeed has 16: I think that pretty much explains why you get some stray decimals in the first case, but not in the second. I guess what we previously had in place to generate those strings did not care much about rounding errors.

If those values were obtained with a previous version of OMC, I'm afraid there's not much we can do, if you don't like 1.000000000000002 you'll have to edit it and save it again. The important point is that decimal numbers _generated by the new Ryu-based infrastructure_ (i.e., what you get by correcting the number and saving it) have enough digits to make sure you get a round number when you later load them again and display them in non-SI unit. In this case, you should get again a round number.

Can you please check that?

**Comment by abuntrock** [comment]
I hope I understood all your points correctly.
Regarding your general question. We have a desired value in a non SI unit because `rpm` is the default unit in most of our applications. We generated the corresponding value in SI units in different ways. I assume that most of the time me and my colleagues got the value from Dymola.
Dymola for example displays both values `rotVel1` and `rotVel2` as desired (1rpm and 1000rpm) even with the different number of significant decimal digits but this is not of importance here.

For the values with not enough significant digits: I edited and saved them again. 
Numbers _generated by the new Ryu-based infrastrucre_ in the way you described above, works fine.

So for me the only question remains, why OMEdit displays `rotVel1` as `0.10471975511966002` effectively adding three significant digits even though the displayUnit equals the SI unit and the value from the text view could be be dispalyed in the parameter dialog without any conversion. 
In the example, this is not important since I will change the value anyway so that `1 rpm` is displayed but maybe its important in other usecases.

**Comment by abuntrock** [comment]
As an additional remark: 
When entering `1rpm` in the dialog in Dymola to model looks like:
```modelica
model Unnamed3
  DisplayUnitRot displayUnitRot1(rotVel1=0.10471975511966) annotation (Placement(transformation(extent={{-58,-14},{-38,6}})));
end Unnamed3;
```
In OMedit the result looks like:
```modelica
model test
  DisplayUnitRot displayUnitRot(rotVel1 = 0.10471975511965978)  annotation(
    Placement(transformation(origin = {-18, 14}, extent = {{-10, -10}, {10, 10}})));  
annotation(
    uses(Modelica(version = "4.0.0"), Rexroth_BRSL(version = "3.0.0")));
end test;
```
The textual representation in OMEdit has three more digits. 
In both tools the model generated with this tool displays the value correctly after reopening the dialog. The only problem is that the model generated with Dymola will not be displayed as desired in OMEdit. But for this we will have to talk to Dassault, I guess.

**Comment by abuntrock** [comment]
I also found a case where generating the value with the new Ryu based infrastructure does not result in the same value being displayed after reopening the parameter dialog.
When adding `2000` as the value for a parameter of type `Modelica.Units.SI.VolumeFlowRate` with the displayUnit being `l/min` results in the following text view:
```modelica
qN = 0.03333333333333333
```
But in the reopened parameter dialog the value is displayed as  `1999.99...98`. I do not know if this represents a corner case as mentioned in #8860

**Comment by casella** [comment]
I'm not sure why Dymola saves converted values with 14 significant decimal digits, but I guess that's not enough to completely specify a double-precision Real number, which needs 16 digits. No wonder if the back-conversion gives a bad rounding, this is caused by the removal of two significant digits in the first place. Maybe Dymola does some kind of magic rounding algorithm to recover the round figure, but it seems to me a bit bad to throw away 2 digits just like that. There's no need to.

**Comment by casella** [comment]
> ```modelica
> 3333333333333333
> ```

I think this may also be because we are dividing by 0.01666666666666667 (which will have an approximated representation) instead of multiplying by 60. Same as discussed in the last comments of 8860

**Comment by casella** [comment]
BTW, if I am not mistaken, IEEE arithmetic guarantees that a*b/b = a, so if we do the conversions by first dividing and then multiplying by the same factor, wes should get the same number back. Currently we divide by a factor and then divide by its inverse, which is not exactly the same thing.

**Comment by bilderbuchi** [comment]
> BTW, if I am not mistaken, IEEE arithmetic guarantees that a*b/b = a, so if we do the conversions by first dividing and then multiplying by the same factor, wes should get the same number back. Currently we divide by a factor and then divide by its inverse, which is not exactly the same thing.

Given the l.h.s. are two separate operations (with rounding steps most probably involved), and IEE754 floats (afaik) not necessarily being associative, I'd be curious what leads you to believe that guarantee to be fulfilled in general?

**Comment by casella** [comment]
@bilderbuchi I thought I read it in the tutorial you posted about rounding. I actually re-read it multiple times but could not find that specific passage, maybe I'm mistaken. 

I'm not sure how Python implements arithmetics, I tried with C
```C
#include <stdio.h>

int main()
{
    printf("%10.20f\n", 0.1);
    printf("%10.20f\n", (0.1*6)/6);
    return 0;
}
```
and I got 

```
0.10000000000000000555
0.10000000000000001943
```
so it seems I am mistaken.

I guess the best thing we can do is to proceed as suggested in [my comment](https://github.com/OpenModelica/OpenModelica/issues/8860#issuecomment-1955209200) in #8860. That may lead a slight loss of precision in some cases, but should avoid this plague of non-rounded round numbers for good.

**Comment by casella** [comment]
Fixed by #12052

**Comment by abuntrock** [comment]
Looks perfect with new nightly, thanks a lot :)
